Homework4

Kiba23 opened the pull request.

Kiba23 committed:
Init commit.

initial commit with one file.

Kiba23 committed:
First breakpoint.

Just initialized a basic logic. Main methods to add in future.

Kiba23 committed:
Second breakpoint.

Initialized the extension method with delegate logic (addition function). All works good. Addition function works only for the same dimensional matrices.

Kiba23 committed:
Third breakpoint.

Redisegned addition function for matrices with different dimensions. Added some tests in Program class.

Kiba23 committed:
Finished the homework4.

Provided tests in Program class. Added MatrixTracker class, implemented event logic. Added Undo method.

radislavkireev commented on Homework4/DiagonalMatrix/Models/DiagonalMatrix.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    public class DiagonalMatrix<T>
+    {
+        public delegate void ElementHandler(int indexI, int indexJ, T oldValue, T newValue);
```
Requirements #5 

**>  Attention: use
> standard practices for working with events!**

The should be a class inherited from event args. 
https://docs.microsoft.com/en-us/dotnet/csharp/event-pattern or check appropriate lection records

radislavkireev commented on Homework4/DiagonalMatrix/MatrixTracker.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public void Undo()
+        {
+            MatrixToTrack[IndexI, IndexJ] = OldValue;
```
When no element have been changed, but Undo called, then [0,0] element is set to default(T).
Handle that case, restore element only when needed.

radislavkireev commented on Homework4/DiagonalMatrix/Extensions/DiagonalMatrixExtension.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+            T[] newElementsAfterAddition;
+
+            if (matrix1.Size == matrix2.Size) // For the matrices with the same dimension
```
You can drop the lines in if {...} and keep only else {...} 

Consider the pseudo code ( else block, line 25, **>** replaced to **>=**)

```
var biggerMatrix = matrix1.Size >= matrix2.Size ? matrix1 : matrix2;
var smallerMatrix = matrix1.Size < matrix2.Size ? matrix1 : matrix2;
```
With that update, for cycle below will stop immediately and the case of equal matrixes will be handler properly. 
`for (int i = smallerMatrix.Size; i < biggerMatrix.Size; i++)`

Kiba23 committed:
Fixed Homework4 after Radislav's suggestions..

Fixed all the things from Radislav. I also have added the HW3 folder, because I accidently removed it from the branch. All works good. Tests are in Program class.
